File: jdave-examples/src/java/jdave/examples/Stack.java
/*
 * Copyright 2006 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jdave.examples;

import java.util.Collection;
import java.util.Iterator;

/**
 * @author Joni Freeman
 * @author Pekka Enberg
 */
public class Stack<T> implements Collection<T> {
    private static final int UNLIMITED = -1;
    private final int maxSize;
    private java.util.Stack<T> stack = new java.util.Stack<T>();
    
    public Stack() {
        this(UNLIMITED);
    }
    
    public Stack(int maxSize) {
        this.maxSize = maxSize;
    }

    public boolean empty() {
        return stack.isEmpty();
    }

    public boolean full() {
        return maxSize != UNLIMITED && stack.size() == maxSize;
    }

    public T peek() {
        return stack.peek();
    }

    public void push(T element) {
        if (full()) {
            throw new StackOverflowException();
        }
        stack.push(element);
    }

    public boolean add(T o) {
        throw new UnsupportedOperationException();
    }

    public boolean addAll(Collection<? extends T> c) {
        throw new UnsupportedOperationException();
    }

    public void clear() {
        throw new UnsupportedOperationException();
    }

    public boolean contains(Object o) {
        return stack.contains(o);
    }

    public boolean containsAll(Collection<?> c) {
        throw new UnsupportedOperationException();
    }

    public boolean isEmpty() {
        throw new UnsupportedOperationException();
    }

    public Iterator<T> iterator() {
        throw new UnsupportedOperationException();
    }

    public boolean remove(Object o) {
        throw new UnsupportedOperationException();
    }

    public boolean removeAll(Collection<?> c) {
        throw new UnsupportedOperationException();
    }

    public boolean retainAll(Collection<?> c) {
        throw new UnsupportedOperationException();
    }

    public int size() {
        throw new UnsupportedOperationException();
    }

    public Object[] toArray() {
        throw new UnsupportedOperationException();
    }

    public <V> V[] toArray(V[] a) {
        throw new UnsupportedOperationException();
    }
}

File: jdave-scala/src/test/jdave/scala/Stack.scala
/*
 * Copyright 2007 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jdave.scala

class Stack[T](maxSize: Int) extends Iterable[T] {
  val UNLIMITED = -1
  val stack = new collection.mutable.Stack[T]()
  
  def this() = this(-1)
  
  def empty() = stack.isEmpty

  def full() = maxSize != UNLIMITED && stack.size >= maxSize

  def peek() = stack.top

  def push(element: T) {
    if (full()) throw new StackOverflowException    
    stack.push(element)
  }

  def contains(o: T) = stack.contains(o)
  
  def size() = stack.size
  
  def pop() = stack.pop
  
  def elements() = stack.elements
}

class StackOverflowException extends Throwable
